### DISCUSSION TITLE
Cloud Optimised Geotiff in imagesharp

### DISCUSSION THREAD

**Discussion by jmkelly** [SixLabors/ImageSharp#1324] [Ideas]
Heya,

**TLDR; is an implementation for Cloud Optimised Geotiff feasible or even sensible in ImageSharp?**

I’ve done some reading, and from what I can see as far as the Tiff branch, I believe COGs could be supported, but I'm not exactly sure what has/hasn't been implemented for TIFFs, and then how much extra will be required for geotiff then COG to be supported.  Maybe this should be considered as an extension to ImageSharp rather than part of the core, and if so, are the extension points there to enable this?  I also realise that the market for geotiff would be substantially less than tiff, and tiff still isn't merged.

Good overviews and links for what a COG is [http://www.cogeo.org](https://www.cogeo.org/)

Some open source implementations I’ve found for reference

- [https://gdal.org/drivers/raster/cog.html](https://gdal.org/drivers/raster/cog.html)  is documentation from [GDAL](https://github.com/OSGeo/gdal) a spatial industry libary that is the defacto standard for spatial image (raster) handling. I believe it's written in C / C++, and heavily used by most raster based spatial libaries and GIS software.
- [https://github.com/geotiffjs/geotiff.js/](https://github.com/geotiffjs/geotiff.js/) implementation in javascript that can be used either client or server side.
- [https://github.com/mapbox/COGDumper/](https://github.com/mapbox/COGDumper/) readonly implementation in pure python

Couple of other links
- [http://download.osgeo.org/geotiff/spec/geotiff.rtf](http://download.osgeo.org/geotiff/spec/geotiff.rtf) geotiff standard, not sure if it contains the COG yet though
- [https://geoexamples.com/other/2019/02/08/cog-tutorial.html](https://geoexamples.com/other/2019/02/08/cog-tutorial.html) good examples of creating / using COGs from various libaries 

If you think this is something to be considered, I'll create an issue, then work out what I can do.  My selfish motivation (longer term) is so I can avoid using GDAL as I find it painful, yet there are very few alternatives, and I like using C# and .netcore.   Short term, I'd like to be able to serve up (via aspnet core) jpeg images from tiles within large (10s - 100s GB compressed Geotiffs) without having to read the whole image. Similar concept to [https://map.openaerialmap.org/](https://map.openaerialmap.org/)

Thanks very much for your consideration!

**Comment by JimBobSquarePants** [comment]
Hi @jmkelly 

Not ignoring this, just need to set aside the time to fully read up before I can comment. A lot to take in!

**Reply by jmkelly** [reply]
@JimBobSquarePants thanks for the reply, all good mate, its a slow burner

**Reply by JimBobSquarePants** [reply]
I never did manage to get around to reading through the specifications. Life kinda got in the way. 

We've come a long way in our TIFF support so maybe you would like to have a look at our existing infrastructure to see if it's possible to add the required support?
